Scripts/interactable/Locomotion/Gimball.lua

```diff
@@ -25,23 +25,62 @@ function Gimball.server_init( self )
 	self.power = 0
 	self.smode = 0
 	self.direction = sm.vec3.new(0,0,1)
+
+	mp_fuel_initialize(self, obj_consumable_gas, 0.35)
 	
 	local stored = self.storage:load()
 	if stored then
-		self.smode = stored - 1
+		local stored_type = type(stored)
+		if stored_type == "number" then
+			self.smode = stored - 1
+		elseif stored_type == "table" then
+			self.smode = stored[1] - 1
+			self.sv_fuel_points = stored[2]
+		end
 	end
+
+	self.sv_saved_fuel_points = self.sv_fuel_points
 end
 
 function Gimball.server_onRefresh( self )
 	self:server_init()
 end
 
-function Gimball.server_onFixedUpdate( self )
-	if self.power ~= 0 and math.abs(self.power) ~= math.huge then
+function Gimball.server_onFixedUpdate( self, dt )
+	local l_container = mp_fuel_getValidFuelContainer(self)
+	local can_activate, can_consume = mp_fuel_canConsumeFuel(self, l_container)
+
+	if can_activate and self.power ~= 0 and math.abs(self.power) ~= math.huge then
 		sm.physics.applyImpulse(self.shape, self.direction*math.abs(self.power), true)
-		--print(self.direction)
+
+		if can_consume then
+			mp_fuel_consumeFuelPoints(self, l_container, self.power, dt)
+		end
+	end
+
+	if self.sv_saved_can_activate ~= can_activate then
+		self.sv_saved_can_activate = can_activate
+		self.network:setClientData(can_activate)
 	end
+
+	if self.sv_saved_fuel_points ~= self.sv_fuel_points then
+		self.sv_saved_fuel_points = self.sv_fuel_points
+		self.storage:save({ self.smode+1, self.sv_saved_fuel_points })
```

Comment:
This will call `self.storage:save` every tick, which is very slow. Maybe it should only be saved once the current fuel points is some amount of points away from the saved fuel points?

Reply:
Fixed